Add struct marshalling design doc.

jkoritzinsky opened the pull request:
Add a design doc explaining the approach we've discussed for struct marshalling.

jkotas commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,127 @@
+
+public class NativeStructMarshallingAttribute : Attribute
+{
+     public NativeStructMarshallingAttribute(Type nativeStructType, string managedToNativeMarshalMethod, string nativeToManagedMarshalMethod, string nativeFreeMethod) {}
```
Is it worth the complexity to make the method names configurable? Can the method names follow a convention instead?

jkoritzinsky replied:
As long as we make the convention well documented so people writing manual marshalling code can easily opt in to this system (and validate that they have it set up correctly via an analyzer), I am totally fine making them follow a convention.

AaronRobinsonMSFT replied:
I would prefer to have an interface that needs to be implemented here or perhaps some kind of registration of function pointer contract.

>Is it worth the complexity to make the method names configurable?

What complexity do you foresee with this approach?

jkotas replied:
> I would prefer to have an interface that needs to be implemented here or perhaps some kind of registration of function pointer contract.

Both interfaces or function pointers are extra indirections. I think they would just get into way for generating efficient marshalers.

> What complexity do you foresee with this approach?

The configurable names make the sources more verbose, but I do not see the value that they add.

AaronRobinsonMSFT replied:
The extra indirections are a concern for me as well. I have measured the `ICustomMarshaler` overheard before and am concerned about how we could improve that. I am also conflicted by success of use. I know this is an advanced feature so that argument is fair. The other side of that is longevity that I want to ensure provides some intuitive sense to adapt and grow without becoming an incrementally designed and free form mess (i.e. support matrix of sadness).

Agreed upon convention is fine but not for naming in my experience because it embeds current thinking and doesn't typically allow much room for change. I don't mind the above in the sense the supplied string represents a method to call that has a defined signature and we insert that string into a format: `"{ret} = {USER_SUPPLIED}({args});"`. If it compiles great, otherwise the supplied function needs to be "fixed".

jkotas replied:
The source generator should be able to inspect methods available on NativeType and pick the best one to call for given scenario; or no method at all if it is not needed.

For example: If `{NativeType}.Free` method exists, the source generator can it to perform cleanup and also generate the associated try/finally exception handling. If `{NativeType}.Free` method does not exist, the source generator can skip generating all that.

AaronRobinsonMSFT replied:
Alright I am becoming less concerned about this. Our support assembly could contain a collection of these "NativeType"s and this would allow interop library authors the flexibility to mix/compose how they want to marshal the managed type. This scales pretty well as we can fold popular "NativeType"s into the support assembly. Is that the path you are seeing here?

jkotas replied:
Yes, this can be one of the paths that this enables.

tannergooding replied:
I don't think simply matching by name alone is good enough. Depending on the exact scenario you may have helper methods or other members available on a type that aren't necessarily meant for use in marshaling.

`IAMDevMemoryAllocator` in particular has a method named `Free()`, as do several other types

tannergooding replied:
I think it needs to be opt-in in some fashion, likely by matching some attribute + signature to ensure that the correct thing always happens.

jkotas replied:
It is opt-in by having dedicated `NativeType` that contains the marshalling methods. This type is different from your mainstream stream, and so there should not be risk of collisions.

jkotas commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,127 @@
+
+The P/Invoke source generator (as well as the struct source generator when nested struct types are used) would use the `BlittableTypeAttribute` and `NativeStructMarshallingAttribute` to determine how to marshal a value type parameter or field instead of looking at the fields of the struct directly.
+
+The largest problem of this design is the switch to an opt-in model for structure types. We could augment this design by providing a fallback to the built-in struct marshaling system via an attribute (either `MarshalAsAttribute` or another one) if we believe providing an explicit escape hatch is needed.
```
This design assumes that there is one and only right way to marshal everything.

Maybe we should allow specifying the native type to use for marshalling in the PInvoke signature. Something like:

```
[DllImport(...)]
static void foo([MarshalUsing(MyNativeType)] MyType a);
```

jkotas replied:
I think the best motivating scenario for this is to allow marshalling of non-blittable types that are owned by somebody else (I have edited my example to reflect it.)

I expect that associated functions would be found by convention on `MyNativeType` in this case.

AaronRobinsonMSFT replied:
Hmmm. With this argument the `BlittableTypeAttribute` type becomes less interesting because if the native type is the managed type that could be interpreted as "Blittable". Right?

jkotas replied:
Agree. `BlittableTypeAttribute` can be encoded as `NativeStructMarshallingAttribute` pointing to self.

jkotas replied:
This may not be a good idea given the method name collisions that it can introduce (see the other comment).

jkoritzinsky replied:
I'd be ok with encoding blittability as a `NativeStructMarshallingAttribute` pointing to self since in the blittable case we wouldn't need to use the marshalling methods (and we can't for cases where we pin).

AaronRobinsonMSFT replied:
>This may not be a good idea given the method name collisions that it can introduce (see the other comment).

Name collisions on what? The supplied "NativeType" would be a new "transport" time simply for the moving between managed/unmanaged.

jkotas replied:
I was referring to https://github.com/dotnet/runtimelab/pull/45#discussion_r465325937.

For example, let's say that we want mark the existing GCHandle type as blittable. 

If it is done via `[BlittableType]`, it is clear what it means. No ambiguity. This type is always passed through without any marshaling.

If it is done via `[NativeStructMarshalling(typeof(GCHandle)]`, it is less clear. For example, should `GCHandle.Free` participate in marshaling or not, assuming `Free` method is part of the convention.

AaronRobinsonMSFT replied:
>If it is done via [NativeStructMarshalling(typeof(GCHandle)], it is less clear. For example, should GCHandle.Free participate in marshaling or not, assuming Free method is part of the convention.

Agreed that is a concern - that was part of my push back when talking about convention. However, it presupposes the need to use existing types as "Transfer types" - I like this term because it describes their purpose. For example, `GCHandle` isn't a transfer type because it already has a purpose and isn't something we would be likely to express on the unmanaged side of the boundary - we can debate that. If it is what we want to reflect on both side, then it is blittable so the collision wouldnt come into play.

Alternatively we could expand the concept mentioned above by merging it with @jkoritzinsky's original proposal.

```csharp
struct MyNameType
{
  // Pre-existing functions on the type
  public static MyNameType Alloc() { ... }
  public static long Free() { ... }

  // Support for DllImport source generator
  // N.B. Signatures are for illustrative purposes only.
  public static IntPtr MyAlloc() { ... }
  public static void MyFree(IntPtr a) { ... }
}

static void foo([MarshalUsing(MyNativeType, Alloc = nameof(MyNameType.MyAlloc), Free = nameof(MyNameType.MyFree))] TypeThatIDoNotOwn a);
```

This allows us by default to have a strongly typed explicit convention or in the cases where it doesn't work the user can say "Use this instead". I genuinely like the merging of the two ideas.

jkotas commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,127 @@
+
+- Usage 1, Source-generated interop:
+
+The user can apply the `GeneratedMarshallingAttribute` to their structure `S`. The source generator will determine if the type is blittable. If it is blittable, the source generator will generate a partial definition and apply the `BlittableTypeAttribute` to the struct type `S`. Otherwise, it will generate a blittable representation of the struct and apply the `NativeStructMarshallingAttribute` a ...
```
Should `NativeStructMarshallingAttribute` be allowed on classes as well? It does not look specific to structs.

AaronRobinsonMSFT replied:
Is the thinking for integrating `ComWrappers` or similar for example?

jkoritzinsky replied:
We can also allow it to be applied to classes similar to how we allow `StructLayoutAttribute` on classes to enable them for interop usage.

For ComWrappers support, we could either use a new attribute, or use Jan's idea for the "transparent struct" concept (if present, a property named `Value` on the native type would be the value we pass down instead of the native type) to implement marshalling via ComWrappers.

For example, something like the following would enable using ComWrappers to marshal a class via this model:

```
[NativeStructMarshalling(typeof(FooNative), ...)]
class Foo
{

}

class FooComWrappers : ComWrappers { ... }

class FooNative
{
      private static readonly FooComWrappers Wrapper = new FooComWrappers();
      public IntPtr Value { get; set; }
     ....
}

```


If we want to take this path to enable ComWrappers, I would suggest that we consider renaming `NativeStructMarshallingAttribute` to `NativeMarshallingAttribute`.

tannergooding commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,127 @@
+
+The user may want to manually mark their types as marshalable in this system due to specific restrictions in their code base around marshaling specific types that the source generator does not account for. We could also use this internally to support custom types in source instead of in the code generator. In this scenario, the user would apply either the `BlittableTypeAttribute` or the `NativeSt ...
+
+The P/Invoke source generator (as well as the struct source generator when nested struct types are used) would use the `BlittableTypeAttribute` and `NativeStructMarshallingAttribute` to determine how to marshal a value type parameter or field instead of looking at the fields of the struct directly.
```
What would happen if the type was marked Blittable but the type contained a non-blittable field?

tannergooding replied:
The two main concerns here are:
1) Users can write their own IL
2) Users can disable analyzer warnings

In both scenarios, you could have `struct S { int _field; }` in the reference assembly but `struct S { bool _x; }` in the implementation.
There are also odd scenarios where the user changes the backing field without a dependent assembly recompiling (which would be a breaking change, but which is still something that may need consideration).

jkoritzinsky replied:
If the user explicitly marks it as blittable, then the analyzer would produce an error. Otherwise, if the type is marked with `[GeneratedMarshalling]`, it will only have the `BlittableTypeAttribute` attribute when it is blittable.

If someone disables the analyzer and applies `BlittableTypeAttribute` manually (which is very much a "don't do this"/unsupported scenario), then we will not generate any marshalling code and behavior at native boundaries would fall into unspecified behavior. (The P/Invoke source generator would currently fall back to the built-in system based on it's implementation)

AaronRobinsonMSFT replied:
>The two main concerns here are:
> 1. Users can write their own IL
> 2. Users can disable analyzer warnings

I think these are fair concerns, but I don't think they change much.

Anyone writing their own IL is in a very advanced scenario and this isn't something I think we can solve or address. That being said, I am very much open for making an attempt additionally I could be misunderstanding the concern here.

Disabling analyzer warnings is perfectly acceptable and that is the prerogative of any developer. If the developer knows better then the compiler/analyzer they are okay going it on their own and if the runtime crashes in an interop call that is on them. Consider the case of "use prior to initialize" in C/C++; if the warning was on the error would have been flagged and could have been corrected. Disabling the warning is on the developer.

The rebuttal to that is the case where .NET should be designed to help users avoid those mistakes and in general I agree. However, this is already an advanced scenario so I would prefer to focus on making analyzers avoid false positives for incorrect use - not noisy - and enable experts to produce high quality AOT friendly interop libraries that users can consume so very few need to actually author P/Invokes - see [dotnet/pinvoke](https://github.com/dotnet/pinvoke).

tannergooding commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,127 @@
+
+The user may want to manually mark their types as marshalable in this system due to specific restrictions in their code base around marshaling specific types that the source generator does not account for. We could also use this internally to support custom types in source instead of in the code generator. In this scenario, the user would apply either the `BlittableTypeAttribute` or the `NativeSt ...
+
+The P/Invoke source generator (as well as the struct source generator when nested struct types are used) would use the `BlittableTypeAttribute` and `NativeStructMarshallingAttribute` to determine how to marshal a value type parameter or field instead of looking at the fields of the struct directly.
```
What is the expected extensibility model, if any, for things which can't currently be surfaced to the runtime today?

For example, the "TransparentStruct" discussion we had previously, where you want to say "I have a struct, but you should pass it like the underlying T instead" because I really just want a convenience API.

or as we start bringing onboard Java (Android) and Objective-C (iOS) support; specific ABI concepts that can't be represented today such as differentiating a C++03 POD type which may have different behavior than a non POD type.

jkotas replied:
The TransparentStruct can be implemented using optional `Value` property on NativeType. If the source generator detected presence of `Value` property on the NativeType, it would use the value of this property for the inner call instead of the NativeType itself. It only works if the type is constant (ie it would not work well for CLong).

The low level ABI concepts would have to be built into the runtime, or we would need to generate a helper C/C++/Objective-C library to implement them on the managed side. I do not think that this proposal helps with that.

tannergooding commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,127 @@
+
+- Usage 1, Source-generated interop:
+
+The user can apply the `GeneratedMarshallingAttribute` to their structure `S`. The source generator will determine if the type is blittable. If it is blittable, the source generator will generate a partial definition and apply the `BlittableTypeAttribute` to the struct type `S`. Otherwise, it will generate a blittable representation of the struct and apply the `NativeStructMarshallingAttribute` a ...
```
For the functions, would there be any "new APIs" for doing low overhead malloc/free like operations? We have `AllocHGlobal` today, but it is fairly expensive (comparatively) and it would be nice to be able to do a simple "allocate native mem", "marshal data", "free native mem"

jkotas replied:
That sounds like an independent proposal / issue to me. `AllocHGlobal` has some unnecessary overhead (in particular on Unix). It is something to fix first.

jkoritzinsky commented:
I've updated the proposal with a few attribute renames and removing the customizability of the native marshalling method names. I've also added a description and a few examples using the "transparent struct" support Jan suggested with the Value property, which actually ended up making this design extremely powerful and versatile while adding minimal complexity.

jkotas commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,225 @@
+     public TNative(TManaged managed)
+     {}
+
+     public TManaged ToManaged() {}
```
Can these methods be optional? E.g. missing `FreeNative` would mean that the type does not own any resources.

jkoritzinsky replied:
I think the `FreeNative` method can definitely be optional. As long as we enforce that at least one of the two marshalling methods (the constructor or the `ToManaged` method) exist, then I'd feel comfortable making all of the methods optional.

jkoritzinsky replied:
And either a missing constructor or a missing ToManaged method would as as a "this direction of marshalling is unsupported" statement.

jkotas commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,225 @@
+
+If the native type `TNative` also has a public `Value` property, then the value of the `Value` property will be passed to native code instead of the `TNative` value itself. As a result, the type `TNative` will be allowed to be non-blittable and the type of the `Value` property will be required to be blittable.
+
+> :question: Should we support opting into marshalling via pinning by detecting a `GetPinnableReference` method with a blittable ref return type as an optional addition to the support of `NativeMarshallingAttribute` and `MarshalUsingAttribute`?
```
> GetPinnableReference

I think so.

Another similar case can be stackalloc buffers. They have to be provided from the outside (unless we invent some new intrinsic way  to allocate buffers on stack).

jkoritzinsky replied:
We would only use stackalloc buffers for types that we would most likely special case in the generator: arrays, strings, and spans. I don't think it would make sense for any user types to opt-in to being allocated via stackalloc instead of just via stack variables.

jkotas replied:
The specialization makes sense for the most performance sensitive types. Once there are data transformations involved, it is slower by design. I do not think it is necessary to special case it for performance. Ideally, it should use regular mechanisms.

For example, it would be nice to be able to marshal strings via a custom encoding (e.g. UTF-32), with the same performance as regular strings.

jkoritzinsky replied:
I'm trying to think of a good way to have a native type opt-in to allocating memory via stackalloc but I'm struggling.

We can't have an `Allocate` method on the native type because that breaks the stackalloc scope.

We could put it on the `NativeMarshalling` or `MarshalUsing` attributes, but then for the UTF-32 string case, it would have to be specified at each usage site.

Any other ideas that I'm thinking of become increasingly obtuse.


Do you have any ideas?

jkotas replied:
I was thinking about something like: If the transport type constructor has a second `Span<char> stackBuffer argument`, the source generator is required to provide stackalloced buffer to it:

```
public struct UTF32StringMarshaler
{
    public UTF32StringMarshaler(string s, Span<char> stackBuffer);

    public const StackBufferSize = 0x100;
}

...
static partial void MyPInvoke([MarshalUsing(UTF32StringMarshaler)] string s);
```

generates:

```
static partial void MyPInvoke(string s)
{
    UTF32Marshaler arg0 = new UTF32StringMarshaler(s, stackalloc char[UTF32Marshaler.StackBufferSize]);
...
}
```

jkotas commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,225 @@
+     public TNative(TManaged managed)
+     {}
+
+     public TManaged ToManaged() {}
```
Nit: `constructor` + `ToManaged` method are asymmetric style-wise. Maybe it should be constructor + conversion operator; or `FromManaged` + `ToManaged` methods.

We can easily support multiple styles if needed.

jkoritzinsky replied:
I'm unsure if I like using a conversion operator because it makes the IDE experience slightly worse and is a little harder to follow. If we go the `FromManaged`/`ToManaged` route, we'd have to decide if we want to use mutable structs and instance methods or use static methods and don't use mutable structures.

jkotas commented on DllImportGenerator/designs/StructMarshalling.md:
````diff
@@ -0,0 +1,247 @@
+}
+```
+
+When these members are both present, the source generator will call the two-parameter constructor with a stack-allocated buffer of `StackBufferSize` bytes when a stack-allocated buffer is usable. As this buffer is guaranteed to be stack allocated and not on the GC heap, it is save to use `Unsafe.AsPointer` to get a pointer to the stack buffer to pass to native code. As a stack-allocated buffer is ...
````
```suggestion
When these members are both present, the source generator will call the two-parameter constructor with a stack-allocated buffer of `StackBufferSize` bytes when a stack-allocated buffer is usable. As this buffer is guaranteed to be stack allocated and not on the GC heap, it is safe to use `Unsafe.AsPointer` to get a pointer to the stack buffer to pass to native code. As a stack-allocated buffer is not usable in all scenarios, for example Reverse P/Invoke and struct marshalling, a one-parameter constructor must also be provided for usage in those scenarios. This may also be provided by providing a two-parameter constructor with a default value for the second parameter.
```

jkotas commented on DllImportGenerator/designs/StructMarshalling.md:
```diff
@@ -0,0 +1,247 @@
+     }
+
+     public HResult ToManaged() => new HResult(Value);
+     public void FreeNative() {}
```
```suggestion
```
